modules/compiler/builtins/abacus/source/abacus_math/ldexp.cpp

```diff
@@ -20,312 +20,121 @@
-    return_value = __abacus_select(x, return_value, isnan);
-  }
-
-  const SignedType is_zero = (x == 0.0f16) & ~is_denorm;
-  SignedType cond4 = __abacus_isinf(x) | isnan | is_zero;
-
-  return __abacus_select(result, return_value, cond4);
-}
-#endif
-
-template <typename T, typename SignedType = typename TypeTraits<T>::SignedType>
-inline T ldexp_float(const T x, const SignedType n) {
-  typedef FPShape<T> Shape;
-
-  // split n into 3 depending on how big it is
-  const abacus_int exp_bias = Shape::Bias();
-  SignedType n_hi = __abacus_select(n, n / 3, __abacus_abs(n) > (exp_bias - 1));
-  SignedType n_lo = __abacus_select(0, n_hi, __abacus_abs(n) > (exp_bias - 1));
-  SignedType n_lo_lo = n - (n_hi + n_lo);
-
-  // put these n's into exponents:
-  n_hi += exp_bias;
-  n_lo += exp_bias;
-  n_lo_lo += exp_bias;
-
-  const SignedType mant_bits = Shape::Mantissa();
-  T mul_hi = abacus::detail::cast::as<T>(n_hi << mant_bits);
-  T mul_lo = abacus::detail::cast::as<T>(n_lo << mant_bits);
-  T mul_lo_lo = abacus::detail::cast::as<T>(n_lo_lo << mant_bits);
-
-  T result = ((x * mul_hi) * mul_lo) * mul_lo_lo;
-
-  // check if n is very big or very small:
-  // The bound is the value of n necessary to go from the smallest number to the
-  // biggest one:
-  //   - 2 * 127: n to go from the smallest normal number to the biggest.
-  //   - 22: n to go from the smallest denormal back into the normal range.
-  //   (smallest denormal is 2 ^ -149, so 149 - 127 = 22).
-  const abacus_int bound = 2 * exp_bias + 22;
-  result = __abacus_select(result, __abacus_copysign(0.0f, x), n < -bound);
-
-  result =
-      __abacus_select(result, __abacus_copysign(ABACUS_INFINITY, x), n > bound);
-
-  // ldexp should still work on denormals, even in FTZ mode, however comparing
-  // a denormal number to 0 in FTZ mode will yield true, so we don't take into
-  // account this comparison if the number is denormal. It also works fine
-  // with denormal support since 0 is not denormal.
-  const SignedType is_denorm = abacus::internal::is_denorm(x);
-  return __abacus_select(
-      result, x,
-      __abacus_isinf(x) | __abacus_isnan(x) | ((x == 0.0f) & ~is_denorm));
+  const SignedType n_c = convert_sat<SignedType>(n);
+
+  // We split n into n_1 and n_2 with n_1 + n_1 + n_2 == n and compute ldexp(x,
+  // n) as ldexp(ldexp(ldexp(x, n_1), n_1), n_2). We need ldexp(ldexp(x, n_1),
+  // n_1) to be exact so that we do not introduce errors due to double rounding,
+  // which requires avoiding turning normals into subnormals early.
+  const SignedType n_1_min =
+      SignedType(1) - (SignedType(((abacus::detail::cast::as<UnsignedType>(x) &
+                                    Shape::ExponentMask()) >>
+                                   Shape::Mantissa())) /
+                       2);
+  const SignedType n_1 = __abacus_max(SignedType(n_c / 3), n_1_min);
+  const SignedType n_2 = n_c - (n_1 << 1);
+
+  const UnsignedType biased_n_1 =
```

Comment:
Is this the bit where it calculates pow(2,n_1) and pow(2,n_2) - I think just an extra line to say what you said before that ldexp -> x * pow(2,n); it just makes it easxier to understand without the full knowledge.

Reply:
I have been working on the assumption that someone working on a standard function would be expected to know what the function is supposed to do, as we tend to not repeat what is already directly in the relevant standards, and that `ldexp(x, n)` returns `x * pow(2, n)` is just its definition, not an implementation detail. I don't think we should single out `ldexp` in documenting things like that, but I don't mind if we want to change that as a general thing across Abacus. That said, agreed that the fact that these lines calculate `pow(2, n_1)` and `pow(2, n_2)` is not obvious, will add a comment about that too.

Reply:
I appreciate this is a standard method, but it does make things easier to review etc, and allows for a quick understanding when there is an issue.  I agree that's part of it's definition though.

Reply:
We talked about this off of GitHub, we do have that documentation at https://github.com/codeplaysoftware/oneapi-construction-kit/blob/f6e8b035978f25ff25cf265bc999104cadafb50d/modules/compiler/builtins/abacus/generate/abacus_math/main.in#L1014-L1022 but it's not in an easily accessible place. We'll work on improving that later.